build_runner/test/server/serve_integration_test.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:build/build.dart';
+import 'package:build_runner/build_runner.dart';
+import 'package:logging/logging.dart';
+import 'package:path/path.dart' as p;
+import 'package:shelf/shelf.dart';
+import 'package:test/test.dart';
+import 'package:watcher/watcher.dart';
+
+import '../common/common.dart';
+
+void main() {
+  FutureOr<Response> Function(Request) handler;
+  InMemoryRunnerAssetReader reader;
+  InMemoryRunnerAssetWriter writer;
+  StreamSubscription subscription;
+  Completer<Null> nextBuild;
+
+  final path = p.absolute('example');
+
+  setUp(() async {
+    final graph = new PackageGraph.fromRoot(
+      new PackageNode.noPubspec('example', path: path),
+    );
+    reader = new InMemoryRunnerAssetReader({}, 'example');
+    writer = new InMemoryRunnerAssetWriter();
+    final server = (await watch(
+      [
+        new BuildAction(const UppercaseBuilder(), 'example'),
+      ],
+      packageGraph: graph,
+      reader: reader,
+      writer: writer,
+      logLevel: Level.ALL,
+      directoryWatcherFactory: (path) => new FakeWatcher(path),
+    ));
+    handler = server.handlerFor('web');
+
+    nextBuild = new Completer<Null>();
+    subscription = server.buildResults.listen((_) {
+      if (!nextBuild.isCompleted) {
+        nextBuild.complete();
```

Comment:
should this just assign `nextBuild` to a `new Completer<Null>()` here? That would probably make us less likely to use it incorrectly if we add more tests here :).